src/utils/taint.js

```diff
@@ -0,0 +1,398 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { CompilerDiagnostic, FatalError } from "../errors.js";
+import type { Realm } from "../realm.js";
+import type { Descriptor, PropertyBinding, ObjectKind } from "../types.js";
+import {
+  DeclarativeEnvironmentRecord,
+  FunctionEnvironmentRecord,
+  ObjectEnvironmentRecord,
+  GlobalEnvironmentRecord,
+} from "../environment.js";
+import {
+  BoundFunctionValue,
+  ProxyValue,
+  SymbolValue,
+  AbstractValue,
+  EmptyValue,
+  FunctionValue,
+  PrimitiveValue,
+  Value,
+  ObjectValue,
+  NativeFunctionValue,
+} from "../values/index.js";
+import { IsDataDescriptor, TestIntegrityLevel } from "../methods/index.js";
+import type { BabelNodeExpression } from "babel-types";
+import invariant from "../invariant.js";
+
+import { ResidualHeapInspector } from "../serializer/ResidualHeapInspector.js";
+
+function emitAllProperties(realm: Realm, O: ObjectValue) {
+  invariant(realm.generator, "a generator must exist while tracking a pure function");
+  let generator = realm.generator;
+  invariant(
+    O !== realm.$GlobalObject,
+    "we should not be able to get here because the global object is never mutated in a pure function."
+  );
+  // Temporary hack
+  let insp = new ResidualHeapInspector(realm, (null: any));
+  for (let [P, propertyBinding] of O.properties) {
+    if (insp.canIgnoreProperty(O, P)) {
+      continue;
+    }
+    let desc = propertyBinding.descriptor;
+    if (desc === undefined) continue; // deleted
+    if (IsDataDescriptor(realm, desc) && desc.configurable && desc.enumerable && desc.writable) {
+      let descValue = desc.value || realm.intrinsics.undefined;
+      invariant(descValue instanceof Value);
+      generator.emitPropertyAssignment(O, P, descValue);
+    } else {
+      generator.emitDefineProperty(O, P, desc);
+    }
+  }
+  // TODO: Emit symbol properties, prototype and other fields as well.
+}
+
+class TaintedValueVisitor {
+  realm: Realm;
+  // Values to visit if they're reachable.
```

Comment:
The comment says "Value", the code says "ObjectValue". Perhaps they should agreed.